File: Blind 75/03 - Sliding Window/76-minimum-window-substring.py
'''
Given two strings s and t of lengths m and n respectively, return the minimum window substring of s such that every character in t (including duplicates) is included in the window. If there is no such substring, return the empty string '.

The testcases will be generated such that the answer is unique.

A substring is a contiguous sequence of characters within the string.
'''

from collections import Counter
from time import time


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        # Initialize overall count map with t
        count_map = Counter(t)
        required = sum(count_map.values())
        start, end = 0, 0
        l = 0
        for r, c in enumerate(s, 1):
            if count_map[c] > 0:
                required -= 1

            # Can be negative (character is not required)
            count_map[c] -= 1

            # Window contains all characters, reduce window to minimum
            if required == 0:
                # Shrink until we reach required character
                while l < r and count_map[s[l]] < 0:
                    count_map[s[l]] += 1
                    l += 1

                # Update window size if smaller than existing
                if end == 0 or r - l < end - start:
                    start, end = l, r

                # Moving right one more loses a required character
                count_map[s[l]] += 1
                required += 1
                l += 1

        return s[start:end]

    def reference(self, s: str, t: str) -> str:
        if t == '':
            return ''

        countT, window = {}, {}
        for c in t:
            countT[c] = 1 + countT.get(c, 0)

        have, need = 0, len(countT)
        res, resLen = [-1, -1], float('infinity')
        l = 0
        for r in range(len(s)):
            c = s[r]
            window[c] = 1 + window.get(c, 0)

            if c in countT and window[c] == countT[c]:
                have += 1

            while have == need:
                # Update our result
                if (r - l + 1) < resLen:
                    res = [l, r]
                    resLen = r - l + 1
                # Pop from the left of our window
                window[s[l]] -= 1
                if s[l] in countT and window[s[l]] < countT[s[l]]:
                    have -= 1
                l += 1

        l, r = res
        return s[l : r + 1] if resLen != float('infinity') else ''

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.minWindow(*case))
                else:
                    self.minWindow(*case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(*case))
                else:
                    self.reference(*case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [('ADOBECODEBANC', 'ABC'), ('a', 'a'), ('a', 'aa')]
    test.quantify(test_cases)

File: Blind 75/03 - Sliding Window/JavaScript/76-minimum-window-substring.js
const { performance } = require("perf_hooks");

class Solution {
  method(s, t) {
    if (s.length < t.length) return "";

    // To store frequency of characters in t
    const charCount = new Array(128).fill(0);
    for (const c of t) {
      charCount[c.charCodeAt(0)]++;
    }

    let required = t.length;
    let l = 0;
    let [minL, minLength] = [0, Infinity];

    for (let r = 0; r < s.length; r++) {
      // Expand the window by including s[r]
      const indexR = s[r].charCodeAt(0);

      // Decrease the required count if the character is in t
      if (charCount[indexR] > 0) required--;
      charCount[indexR]--;

      // Once all characters are in the window, try to shrink it
      while (required === 0) {
        const indexL = s[l].charCodeAt(0);

        // Update the minimum window if this is smaller
        if (r - l + 1 < minLength) {
          minLength = r - l + 1;
          minL = l;
        }

        // Shrink the window by moving left pointer
        charCount[indexL]++;
        if (charCount[indexL] > 0) required++;
        l++;
      }
    }

    return minLength === Infinity ? "" : s.substring(minL, minL + minLength);
  }

  reference(s, t) {
    const isMissingArgs = !s.length || !t.length;
    if (isMissingArgs) return "";

    const frequencyMap = this.getFrequencyMap(t);
    const { start, end } = this.getWindowPointers(s, t, frequencyMap);

    return this.getSubString(s, start, end);
  }

  getFrequencyMap = (str, frequencyMap = new Map()) => {
    for (const char of str) {
      frequencyMap.set(char, (frequencyMap.get(char) || 0) + 1);
    }

    return frequencyMap;
  };

  getWindowPointers = (s, t, frequencyMap) => {
    let [left, right, matched, start, end] = [0, 0, 0, 0, s.length + 1];

    while (right < s.length) {
      matched = this.addRightFrequency(s, right, frequencyMap, matched);

      const canSlide = () => matched === t.length;
      while (canSlide()) {
        const window = right - left + 1;

        const isSmaller = window < end;
        if (isSmaller) {
          [start, end] = [left, window];
        }

        matched = this.subtractLeftFrequency(s, left, frequencyMap, matched);
        left++;
      }

      right++;
    }

    return { start, end };
  };

  addRightFrequency = (s, right, frequencyMap, matched) => {
    const char = s[right];

    if (frequencyMap.has(char)) {
      frequencyMap.set(char, frequencyMap.get(char) - 1);

      const isInWindow = 0 <= frequencyMap.get(char);
      if (isInWindow) matched++;
    }

    return matched;
  };

  subtractLeftFrequency = (s, left, frequencyMap, matched) => {
    const char = s[left];

    if (frequencyMap.has(char)) {
      const isOutOfWindow = frequencyMap.get(char) === 0;
      if (isOutOfWindow) matched--;

      frequencyMap.set(char, frequencyMap.get(char) + 1);
    }

    return matched;
  };

  getSubString = (s, start, end) =>
    end <= s.length ? s.slice(start, start + end) : "";

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(...input));
        else this.method(...input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(...input));
        else this.reference(...input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  ["ADOBECODEBANC", "ABC"],
  ["a", "a"],
  ["a", "aa"],
];
test.quantify(testCases);
